modules/renter/renter.go

```diff
@@ -123,36 +122,110 @@ func newRenter(cs modules.ConsensusSet, tpool modules.TransactionPool, hdb hostD
 	if tpool == nil {
 		return nil, errNilTpool
 	}
+	if hc == nil {
+		return nil, errNilContractor
+	}
 	if hdb == nil {
 		// Nil hdb currently allowed for testing purposes. :(
 		// return nil, errNilHdb
 	}
 
 	r := &Renter{
-		cs:             cs,
-		hostDB:         hdb,
-		hostContractor: hc,
-
+		newFiles: make(chan *file),
 		files:    make(map[string]*file),
 		tracking: make(map[string]trackedFile),
 
-		persistDir: persistDir,
-		mu:         sync.New(modules.SafeMutexDelay, 1),
+		workerPool: make(map[types.FileContractID]*worker),
+
+		cs:             cs,
+		hostDB:         hdb,
+		hostContractor: hc,
+		persistDir:     persistDir,
+		mu:             sync.New(modules.SafeMutexDelay, 1),
+		tg:             new(sync.ThreadGroup),
 	}
 	if err := r.initPersist(); err != nil {
 		return nil, err
 	}
 
+	// Spin up the workers for the work pool.
+	contracts := hc.Contracts()
+	for _, c := range contracts {
+		err := r.addWorker(c.ID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
 	go r.threadedRepairLoop()
 
 	return r, nil
 }
 
 // Close closes the Renter and its dependencies
 func (r *Renter) Close() error {
+	r.tg.Stop()
 	return r.hostDB.Close()
 }
 
+// SetSettings will update the settings for the renter.
+func (r *Renter) SetSettings(s modules.RenterSettings) error {
+	err := r.hostContractor.SetAllowance(s.Allowance)
+	if err != nil {
+		return err
+	}
+
+	id := r.mu.Lock()
+	defer r.mu.Unlock(id)
+
+	// Get the list of contracts in the contractor.
+	newContracts := r.hostContractor.Contracts()
+
+	// Get the list of contracts in the work pool.
+	oldContracts := make([]types.FileContractID, 0, len(r.workerPool))
+	for fcid := range r.workerPool {
+		oldContracts = append(oldContracts, fcid)
+	}
+
+	// Find all elements in newContracts that are not in oldContracts and make
+	// workers for them.
+	for _, nc := range newContracts {
+		found := false
+		for _, ofcid := range oldContracts {
+			if nc.ID == ofcid {
+				found = true
+				break
+			}
+		}
+		if !found {
+			err = r.addWorker(nc.ID)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Find all elements in oldContracts that are not in newContracts and kill
+	// their workers.
```

Comment:
I think this can be cleaned up by using maps:
```go
newContracts := make(map[types.FilecontractID]struct{})
for _, id := range r.hostContractor.Contracts() {
    newContracts[id] = struct{}{}
}
oldContracts := r.workerPool // just need ID

// add a worker for each newContract that is not in oldContracts
for id := range newContracts {
    if _, ok := oldContracts[id]; !ok {
        r.addWorker(id)
    }
}
// remove any workers that are in oldContracts but not newContracts
for id := range oldContracts {
    if _, ok := newContracts[id]; !ok {
        r.retireWorker(id)
    }
}
```